dali/pipeline/data/tensor_vector.cc

```diff
@@ -576,191 +686,256 @@ bool TensorVector<Backend>::is_pinned() const {
+  contiguous_buffer_.set_device_id(device_id);
   for (auto &t : tensors_) {
-    t->set_device_id(device_id);
+    t.set_device_id(device_id);
   }
+  device_ = device_id;
 }
 
 
 template <typename Backend>
 int TensorVector<Backend>::device_id() const {
-  if (IsContiguous()) {
-    return tl_->device_id();
-  } else if (!tensors_.empty()) {
-    return tensors_[0]->device_id();
-  }
-  return CPU_ONLY_DEVICE_ID;
+  return device_;
 }
 
 
 template <typename Backend>
 void TensorVector<Backend>::reserve(size_t total_bytes) {
-  if (state_ == State::noncontiguous) {
+  int batch_size_bkp = curr_num_tensors_;
+  if (!state_.IsContiguous()) {
     tensors_.clear();
-    curr_num_tensors_ = 0;
+    resize_tensors(0);
+  }
+  state_.Setup(BatchState::Contiguous);
+  contiguous_buffer_.reserve(total_bytes);
+  if (IsValidType(type_)) {
+    resize_tensors(batch_size_bkp);
+    recreate_views();
   }
-  state_ = State::contiguous;
-  tl_->reserve(total_bytes);
-  UpdateViews();
 }
 
 
 template <typename Backend>
 void TensorVector<Backend>::reserve(size_t bytes_per_sample, int batch_size) {
   assert(batch_size > 0);
-  state_ = State::noncontiguous;
+  state_.Setup(BatchState::Noncontiguous);
   resize_tensors(batch_size);
   for (int i = 0; i < curr_num_tensors_; i++) {
-    tensors_[i]->reserve(bytes_per_sample);
+    tensors_[i].reserve(bytes_per_sample);
   }
 }
 
 
 template <typename Backend>
 bool TensorVector<Backend>::IsContiguous() const noexcept {
-  return state_ == State::contiguous;
+  return state_.IsContiguous();
 }
 
 
 template <typename Backend>
-void TensorVector<Backend>::SetContiguous(bool contiguous) {
-  if (contiguous) {
-    state_ = State::contiguous;
-  } else {
-    state_ = State::noncontiguous;
+void TensorVector<Backend>::recreate_views() {
+  // precondition: type, shape are configured
+  uint8_t *sample_ptr = static_cast<uint8_t *>(contiguous_buffer_.raw_mutable_data());
+  int64_t num_samples = shape().num_samples();
+  for (int64_t i = 0; i < num_samples; i++) {
+    // or any other way
+    auto tensor_size = shape().tensor_size(i);
+
+    std::shared_ptr<void> sample_alias(contiguous_buffer_.get_data_ptr(), sample_ptr);
+    tensors_[i].ShareData(sample_alias, tensor_size * type_info().size(), is_pinned(), shape()[i],
+                          type(), device_id(), order());
+    tensors_[i].set_device_id(device_id());
+    sample_ptr += tensor_size * type_info().size();
+  }
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::SetBatchState(BatchState state) {
+  if (state == BatchState::Default) {
+    // remove the force, keep the current state information
+    state_.Setup(state_.Get(), false);
+    return;
+  }
+  DALI_ENFORCE(state_.Get() == state || !has_data(),
+               "Contiguous or non-contiguous mode cannot be set to already allocated buffer.");
+  state_.Setup(state, true);
+}
+
+
+template <typename Backend>
+void TensorVector<Backend>::SetBatchState(bool contiguous) {
```

Comment:
We shouldn't really have such method: `tv.SetBatchState(true)` is very misleading to read. Do we really need bool and `enum BatchState` versions?

Reply:
Initially I didn't have them, than I thought they can be kinda useful, but I agree, that we should probably strive for more type safety and having an explicit thing to put here is better. Removing.